This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a binary classification problem that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks) Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv("./sonar.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
110  0.0210  0.0121  0.0203  0.1036  0.1675  0.0418  0.0723  0.0828  0.0494   
173  0.0329  0.0216  0.0386  0.0627  0.1158  0.1482  0.2054  0.1605  0.2532   
172  0.0180  0.0444  0.0476  0.0698  0.1615  0.0887  0.0596  0.1071  0.3175   
88   0.0274  0.0242  0.0621  0.0560  0.1129  0.0973  0.1823  0.1745  0.1440   
30   0.0240  0.0218  0.0324  0.0569  0.0330  0.0513  0.0897  0.0713  0.0569   

         9   ...      51      52      53      54      55      56      57  \
110  0.0686  ...  0.0117  0.0101  0.0061  0.0031  0.0099  0.0080  0.0107   
173  0.2672  ...  0.0095  0.0151  0.0059  0.0015  0.0053  0.0016  0.0042   
172  0.2918  ...  0.0122  0.0114  0.0098  0.0027  0.0025  0.0026  0.0050   
88   0.1808  ...  0.0113  0.0108  0.0085  0.0047  0.0074  0.0104  0.0161   
30   0.0389  ...  0.0162  0.0146  0.0093  0.0112  0.0094  0.0054  0.0019   

         58      59  60  
110  0.0161  0.0133   M  
173  0.0053  0.0074   M  
172  0.0073  0.0022   M  
88   0.0220  0.0173   R  
30   0.0066  0.0023   R  

[5 rows x 61 columns]

In [4]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [5]:
df[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

In [7]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [8]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
107  0
41   1
189  0
182  0
1    1

In [9]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [10]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

## Model without Dropout Layer

In [12]:
import tensorflow as tf
from tensorflow import keras

In [13]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6936 - accuracy: 0.5532
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6819 - accuracy: 0.5573
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6496 - accuracy: 0.6440
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6432 - accuracy: 0.7235
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6108 - accuracy: 0.7270
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5731 - accuracy: 0.7439
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5501 - accuracy: 0.8033
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4976 - accuracy: 0.7617
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4268 - accuracy: 0.8492
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4650 - accuracy: 0.8244
Epoch 11/

20/20 [==============================] - 0s 4ms/step - loss: 0.0279 - accuracy: 1.0000
Epoch 83/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0192 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0219 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0184 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0137 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0119 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0198 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0173 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0125 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0118 - accuracy: 1.0000
Epoch 92/100


In [14]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 0s/step - loss: 0.7492 - accuracy: 0.8077


[0.7492457628250122, 0.807692289352417]

In [15]:
y_pred = model.predict(X_test).reshape(-1)

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[0. 0. 1. 0. 1. 1. 0. 1. 0. 1.]


In [16]:
from sklearn.metrics import confusion_matrix , classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.74      0.96      0.84        27
           1       0.94      0.64      0.76        25

    accuracy                           0.81        52
   macro avg       0.84      0.80      0.80        52
weighted avg       0.84      0.81      0.80        52



## Model with Dropout Layer

In [17]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7758 - accuracy: 0.4752
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7029 - accuracy: 0.5617
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7074 - accuracy: 0.5411
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7064 - accuracy: 0.5114
Epoch 5/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6949 - accuracy: 0.5037
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7037 - accuracy: 0.4735
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6744 - accuracy: 0.6009
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6981 - accuracy: 0.4978
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6904 - accuracy: 0.4967
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7083 - accuracy: 0.4742
Epoch 11/

20/20 [==============================] - 0s 2ms/step - loss: 0.3116 - accuracy: 0.8826
Epoch 82/100
20/20 [==============================] - 0s 3ms/step - loss: 0.4380 - accuracy: 0.8224
Epoch 83/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3317 - accuracy: 0.8779
Epoch 84/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3049 - accuracy: 0.9102
Epoch 85/100
20/20 [==============================] - ETA: 0s - loss: 0.4467 - accuracy: 0.75 - 0s 2ms/step - loss: 0.3725 - accuracy: 0.8093
Epoch 86/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3501 - accuracy: 0.8894
Epoch 87/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3714 - accuracy: 0.8650
Epoch 88/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3049 - accuracy: 0.8957
Epoch 89/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3574 - accuracy: 0.8718
Epoch 90/100
20/20 [==============================] - 0s 2ms/step - los

In [18]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 0s/step - loss: 0.4339 - accuracy: 0.8077


[0.4339114725589752, 0.807692289352417]

In [19]:
y_pred = modeld.predict(X_test).reshape(-1)

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [20]:
from sklearn.metrics import confusion_matrix , classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      0.89      0.83        27
           1       0.86      0.72      0.78        25

    accuracy                           0.81        52
   macro avg       0.82      0.80      0.81        52
weighted avg       0.81      0.81      0.81        52



Using dropout layer test accuracy increased from 0.77 to 0.81